uswds/uswds
# USWDS - Fieldset: Resolve legend margin error

# Summary

Resolved `usa-fieldset` & `usa-legend` margin conflict. `usa-fieldset` now appropriately respects the margin set by child legend elements.

## Breaking change

This is _potentially_ a breaking change.

> **Warning**
> In Chrome and Safari, anywhere there is `usa-legend` nested beneath `usa-fieldset`, a larger margin is rendered than before. This comes from the margin set by `usa-legend` which was hidden by this bug.

## Related issue

Closes #4153 

## Related pull requests

[Form change log item →](https://github.com/uswds/uswds-site/pull/2192)

## Preview link
[Checkbox →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-fieldset-legend-margin/iframe.html?args=&id=components-form-inputs-checkbox--default)
[Mailing Address →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-fieldset-legend-margin/iframe.html?args=&id=patterns-forms--mailing-address)
[Create Account Page →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-fieldset-legend-margin/iframe.html?args=&id=pages-create-account--create-account-page)

## Problem statement

`fieldset` does not respect margins set by `usa-legend` or `usa-legend--large`.


## Solution

Add `display: contents` to `usa-fieldset` styles. 

> [display: contents](https://drafts.csswg.org/css-display/#propdef-display) simply removes their [principal box](https://drafts.csswg.org/css-display/#principal-box), and their contents render as normal. [^1]

[^1]: https://drafts.csswg.org/css-display/#unbox:~:text=display%3A%20contents%20simply%20removes%20their%20principal%20box%2C%20and%20their%20contents%20render%20as%20normal.

## Testing and review
1. Inspect components which use `usa-fieldset` and `usa-legend`. Ex:
   - [Checkbox →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-fieldset-legend-margin/iframe.html?args=&id=components-form-inputs-checkbox--default)
   - [Mailing Address →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-fieldset-legend-margin/iframe.html?args=&id=patterns-forms--mailing-address)
   - [Create Account Page →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-fieldset-legend-margin/iframe.html?args=&id=pages-create-account--create-account-page)
2. Confirm the margin set by `usa-legend` or `usa-legend--large` is respected and visable.

It may be helpful to duplicate the `fieldset` element in the preview and make sure there is a visible margin between them

### Testing Checklist
- [ ] Confirm margin is respected
- [ ] Removing `display: contents` from `usa-fieldset` causes margin to be disregarded
- [ ] Confirm there is no unintended visual regression in components using `usa-fieldset`

## Additional content

### Screenshots from issue
**Current Behavior:**
![Screen Shot 2021-04-14 at 1 47 24 PM](https://user-images.githubusercontent.com/20175921/114777078-16b2c780-9d28-11eb-8e2e-9f04f487c87c.png)
**Expected Behavior:**
![Screen Shot 2021-04-14 at 2 24 45 PM](https://user-images.githubusercontent.com/20175921/114781215-41ebe580-9d2d-11eb-9570-10a63ed28069.png)

---

Before opening this PR, make sure you’ve done whichever of these applies to you:
